// yaml-types/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.symbol = exports.sharedSymbol = exports.regexp = exports.nullobject = exports.functionTag = exports.error = exports.classTag = exports.bigint = void 0;
var bigint_js_1 = require("./bigint.js");
Object.defineProperty(exports, "bigint", { enumerable: true, get: function () { return bigint_js_1.bigint; } });
var class_js_1 = require("./class.js");
Object.defineProperty(exports, "classTag", { enumerable: true, get: function () { return class_js_1.classTag; } });
var error_js_1 = require("./error.js");
Object.defineProperty(exports, "error", { enumerable: true, get: function () { return error_js_1.error; } });
var function_js_1 = require("./function.js");
Object.defineProperty(exports, "functionTag", { enumerable: true, get: function () { return function_js_1.functionTag; } });
var null_object_js_1 = require("./null-object.js");
Object.defineProperty(exports, "nullobject", { enumerable: true, get: function () { return null_object_js_1.nullobject; } });
var regexp_js_1 = require("./regexp.js");
Object.defineProperty(exports, "regexp", { enumerable: true, get: function () { return regexp_js_1.regexp; } });
var symbol_js_1 = require("./symbol.js");
Object.defineProperty(exports, "sharedSymbol", { enumerable: true, get: function () { return symbol_js_1.sharedSymbol; } });
Object.defineProperty(exports, "symbol", { enumerable: true, get: function () { return symbol_js_1.symbol; } });
